controllers/rollingupgrade_controller.go

```diff
@@ -244,28 +248,133 @@ func (r *RollingUpgradeReconciler) CallKubectlDrain(ctx context.Context, nodeNam
+		time.Sleep(time.Duration(ScaleoutSleepIntervalSeconds) * time.Second)
+	}
+	return nil
+}
+
+func (r *RollingUpgradeReconciler) WaitForDesiredNodes(ruObj *upgrademgrv1alpha1.RollingUpgrade) error {
+	var (
+		started = time.Now()
+	)
+	for {
+		if time.Since(started) >= (time.Second * time.Duration(ScaleoutTimeoutSeconds)) {
+			err := errors.New("WaitForDesired timed out while waiting for instance to be added")
+			return err
+		}
+
+		r.populateNodeList(ruObj, r.generatedClient.CoreV1().Nodes())
+
+		// get list of inService instance IDs
+		inServiceInstances := getInServiceIds(r.Asg.Instances)
+		desiredCapacity := aws.Int64Value(r.Asg.DesiredCapacity)
+
+		// check all of them are nodes and are ready
+		var foundCount int64 = 0
+		for _, node := range r.NodeList.Items {
+			tokens := strings.Split(node.Spec.ProviderID, "/")
+			instanceID := tokens[len(tokens)-1]
+			if contains(inServiceInstances, instanceID) && isNodeReady(node) {
+				foundCount++
+			}
+		}
+		if foundCount == desiredCapacity {
+			break
+		}
+
+		log.Printf("%v: new node has not yet joined the cluster, will wait %vs and retry", ruObj.Name, ScaleoutSleepIntervalSeconds)
+		time.Sleep(time.Duration(ScaleoutSleepIntervalSeconds) * time.Second)
+	}
+	return nil
+}
+
+func (r *RollingUpgradeReconciler) WaitForTermination(ruObj *upgrademgrv1alpha1.RollingUpgrade, nodeName string, nodeInterface v1.NodeInterface) (bool, error) {
 	var (
 		started = time.Now()
 	)
 	for {
 		if time.Since(started) >= (time.Second * time.Duration(TerminationTimeoutSeconds)) {
-			log.Println("WaitForTermination timed out while waiting for node to unjoin")
+			log.Printf("%v: WaitForTermination timed out while waiting for node to unjoin", ruObj.Name)
 			return false, nil
 		}
 
 		_, err := nodeInterface.Get(nodeName, metav1.GetOptions{})
 		if v1errors.IsNotFound(err) {
-			log.Printf("node %s is unjoined from cluster, upgrade will proceed", nodeName)
+			log.Printf("%v: node %s is unjoined from cluster, upgrade will proceed", ruObj.Name, nodeName)
 			break
 		}
 
-		log.Printf("node %s is still joined to clutster, will wait %vs and retry", nodeName, TerminationSleepIntervalSeconds)
+		log.Printf("%v: node %s is still joined to clutster, will wait %vs and retry", ruObj.Name, nodeName, TerminationSleepIntervalSeconds)
 		time.Sleep(time.Duration(TerminationSleepIntervalSeconds) * time.Second)
 	}
 	return true, nil
 }
 
+// SetStandby sets the autoscaling instance to standby mode.
+func (r *RollingUpgradeReconciler) SetStandby(ruObj *upgrademgrv1alpha1.RollingUpgrade, instanceID string) error {
+
+	log.Infof("%v: Setting %v to stand-by", ruObj.Name, instanceID)
+	input := &autoscaling.EnterStandbyInput{
+		AutoScalingGroupName:           aws.String(ruObj.Spec.AsgName),
+		InstanceIds:                    aws.StringSlice([]string{instanceID}),
+		ShouldDecrementDesiredCapacity: aws.Bool(false),
+	}
+
+	err := r.populateAsg(ruObj)
```

Comment:
Is this call to `populateASG` needed? It will make AWS API calls and might get throttled, etc.

`populateAsg` already stores the result it got from AWS into `ruObj`. Should we just use that?

Reply:
Unfortunately we have to call this for waiters.
Otherwise we will never detect a state has changed.

For example, we are waiting for N instances to be in-service, we must keep refreshing the "cached" version.

In this case (in SetStandby) not sure what was the reason for adding this here.
I will test without it and see if it can be removed